openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1577 @@
+#endif
+
+        mParent->mSweepMask.clear();
+        mParent->mSweepMask.topologyUnion(mParent->mSdfGrid->constTree());
+
+        using LeafManagerT = tree::LeafManager<SweepMaskTreeT>;
+        using LeafT = typename SweepMaskTreeT::LeafNodeType;
+        LeafManagerT leafManager(mParent->mSweepMask);
+
+        auto kernel = [&](LeafT& leaf, size_t /*leafIdx*/) {
+            static const SdfValueT Unknown = std::numeric_limits<SdfValueT>::max();
+            SdfAccT acc(mSdfGrid->tree());
+            // The following hack is safe due to the topoloyUnion in
+            // init and the fact that SdfValueT is known to be a floating point!
+            SdfValueT *data = acc.probeLeaf(leaf.origin())->buffer().data();
+            for (auto voxelIter = leaf.beginValueOn(); voxelIter; ++voxelIter) {// mask voxels
+                if (math::Abs( data[voxelIter.pos()] ) < Unknown ) {
+                    // disable boundary voxels from the mask tree
+                    voxelIter.setValue(false);
+                }
+            }
+        };
+        leafManager.foreach( kernel );
+
+        // cache the leaf node origins for fast lookup in the sweeping kernels
+        mParent->computeSweepMaskLeafOrigins();
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.stop();
+#endif
+    }// FastSweeping::MaskKernel::run
+
+    // Private member data of MaskKernel
+    FastSweeping *mParent;
+    SdfGridT     *mSdfGrid;//raw pointer, i.e. lock free
+};// FastSweeping::MaskKernel
+
+/// @brief Private class of FastSweeping to perform concurrent fast sweeping in two directions
+template <typename SdfGridT, typename ExtValueT>
+struct FastSweeping<SdfGridT, ExtValueT>::SweepingKernel
+{
+    SweepingKernel(FastSweeping &parent) : mParent(&parent) {}
+    SweepingKernel(const SweepingKernel&) = delete;
+    SweepingKernel& operator=(const SweepingKernel&) = delete;
+
+    /// Main method that performs concurrent bi-directional sweeps
+    template<typename HashOp>
+    void computeVoxelSlices(HashOp hash)
+    {
+#ifdef BENCHMARK_FAST_SWEEPING
+        util::CpuTimer timer;
+#endif
+
+        // mask of the active voxels to be solved for, i.e. excluding boundary voxels
+        const SweepMaskTreeT& maskTree = mParent->mSweepMask;
+
+        using LeafManagerT = typename tree::LeafManager<const SweepMaskTreeT>;
+        using LeafT = typename SweepMaskTreeT::LeafNodeType;
+        LeafManagerT leafManager(maskTree);
+
+        // compute the leaf node slices that have active voxels in them
+        // the sliding window of the has keys is -14 to 21 (based on an 8x8x8 leaf node
+        // and the extrema hash values i-j-k and  i+j+k), but we use a larger mask window here to
+        // easily accomodate any leaf dimension. The mask offset is used to be able to
+        // store this in a fixed-size byte array
+        constexpr int maskOffset = LeafT::DIM * 3;
+        constexpr int maskRange = maskOffset * 2;
```

Comment:
Shouldn't the maskOffset be `(LeafT::DIM-1) * 2` and the maskRange be `maskOffset + ((LeafT::DIM-1) * 3) + 1`? i.e., the lowest possible value is `(LeafT::DIM-1)  * 2` and the largest possible id is `(LeafT::DIM-1)  *  2 + max` where `max` is simply `(LeafT::DIM-1) + (LeafT::DIM-1) + (LeafT::DIM-1)`.

also spl "hash"

Reply:
Dan added this

Reply:
@danrbailey thoughts?

Reply:
This is intentionally an over-estimate. I initially had some more complicated logic in there similar to what you're proposing, but I simplified it to reduce the risk that I'd missed an edge case or that it wouldn't work for a particular leaf node configuration. There's a much greater benefit from using this voxel slicing algorithm than what you would save in this small optimization so I opted for caution at this stage.

In fact, there are even better schemes than what you're proposing if you factor in the exact hash and you compute an offset and range per hash. On top of that, there are more complex but much more memory efficient schemes than storing a nodemask per slice. Finally, there's the improvements we discussed the other day of segmenting the active voxels into different regions that can be computed concurrently. All of these improvements and optimisations won't affect the API in any way, so they can always come later.

Reply:
Thanks @danrbailey, still a bit confused though - for my own understanding, are the values I've calculated correct and a better representation of the required sizes? You mention you've over estimated for a "particular leaf node config", can you give an example? This seems like a pretty arbitrary choice, no other algorithm works as such and expects a uniform leaf node dim. In fact this may make the situation worse by working for these other leaf node types and not others! We should at least be asserting here in that case - I don't like the idea that we're just randomly choosing an allocation size to make sure we haven't missed any phantom edge cases :\

To be clear, the only reason I chose to look at these lines in more detail is *because* of your decision to use this particular math. It didn't make any sense to me! It unfortunately only serves to obfuscate the intended implementation to the reader.